### ISSUE TITLE
tf.matmul() fails with ragged inputs of shape [batch_size, None, dims] and transpose_b=True

### ISSUE THREAD

**Issue by ageron** [tensorflow/tensorflow#55209]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Debian 11.2 "bullseye" (keras-dev docker image)
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: N/A
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): v1.12.1-72504-gfeb9095a53f 2.9.0-dev20220311
- Python version: 3.9.10
- Bazel version (if compiling from source): N/A
- GCC/Compiler version (if compiling from source): N/A
- CUDA/cuDNN version: N/A
- GPU model and memory: N/A

**Describe the current behavior**

When A and B are two ragged tensors, both of shape [batch_size, None, dims], then `tf.matmul(A, B, transpose_b=True)` fails. However, when A and B are ragged tensors of shape [batch_size, None, None], with the exact same values, then it works fine.

**Describe the expected behavior**

I expect the first scenario to work, and it should return the same result as the second scenario.

**[Contributing](https://www.tensorflow.org/community/contribute)**

- Do you want to contribute a PR? (yes/no): not sure I'll have time, but if it's not too long ok
- Briefly describe your candidate solution(if contributing): I'm guessing there's a bug in the shape checks in `tf.matmul()` since I see no reason why [batch_size, None, dims] would fail while [batch_size, None, None] would succeed. If anything, I would have expected the opposite.

**Standalone code to reproduce the issue**

See this [gist](https://colab.research.google.com/gist/ageron/1226921dbd2be65dc69f12847f2c229b/tensorflow-issue-matmul-ragged.ipynb).

This code fails:

```python
import tensorflow as tf

A_ids = tf.ragged.constant([[1, 2], [1, 0, 5]])
B_ids = tf.ragged.constant([[2, 4, 6, 9], [3]])
embedding_layer = tf.keras.layers.Embedding(10, 3)
A = embedding_layer(A_ids)  # A has shape [2, None, 3]
B = embedding_layer(B_ids)  # B has shape [2, None, 3]
C = tf.matmul(A, B, transpose_b=True)  # RAISES AN InvalidArgumentError!!!
```

But this one succeeds:

```python
import tensorflow as tf

A = tf.ragged.constant([  # A has shape [2, None, None]
    [[1., 2., 3.], [3., 4., 5.]],
    [[1., 3., 5.], [5., 7., 9.], [9., 11., 13.]]
])
B = tf.ragged.constant([  # B has shape [2, None, None]
    [[10., 20., 30.], [30., 40., 50.], [50., 60., 70.], [70., 80., 90.]],
    [[11., 21., 31.]]
])

C = tf.matmul(A, B, transpose_b=True)  # WORKS FINE NOW!
```

The only difference I can see is that the shape of `A` and `B` is `[2, None, 3]` in the first case, but it's `[2, None, None]` in the second case. It doesn't make sense.

**Other info / logs**

Below is the full stacktrace:

<details>

```stacktrace
---------------------------------------------------------------------------
InvalidArgumentError                      Traceback (most recent call last)
[<ipython-input-8-3e4e536edcee>](https://localhost:8080/#) in <module>()
      6 A = embedding_layer(A_ids)  # A has shape [2, None, 3]
      7 B = embedding_layer(B_ids)  # B has shape [2, None, 3]
----> 8 C = tf.matmul(A, B, transpose_b=True)  # ERROR!

1 frames
[/usr/local/lib/python3.7/dist-packages/tensorflow/python/util/traceback_utils.py](https://localhost:8080/#) in error_handler(*args, **kwargs)
    151     except Exception as e:
    152       filtered_tb = _process_traceback_frames(e.__traceback__)
--> 153       raise e.with_traceback(filtered_tb) from None
    154     finally:
    155       del filtered_tb

[/usr/local/lib/python3.7/dist-packages/tensorflow/python/framework/ops.py](https://localhost:8080/#) in raise_from_not_ok_status(e, name)
   7184 def raise_from_not_ok_status(e, name):
   7185   e.message += (" name: " + name if name is not None else "")
-> 7186   raise core._status_to_exception(e) from None  # pylint: disable=protected-access
   7187 
   7188 

InvalidArgumentError: All flat_values must have compatible shapes.  Shape at index 0: [4].  Shape at index 1: [1].  If you are using tf.map_fn, then you may need to specify an explicit fn_output_signature with appropriate ragged_rank, and/or convert output tensors to RaggedTensors. [Op:RaggedTensorFromVariant]
```

</details>

**Related issues**

* This TF issue is probably the cause of this Keras issue: https://github.com/keras-team/keras/issues/16226
* There's a closed issue about TF matmul + ragged tensors, but I think it's different: https://github.com/tensorflow/tensorflow/issues/28109

**Comment by ageron** [comment]
Until this issue is fixed, here's a workaround: the following function sets the last dimension to `None`, so the `matmul()` operation succeeds.

```python
@tf.function
def unset_outer_dimension(ragged_3d):
  row_lengths = ragged_3d.row_lengths()
  total_steps = tf.reduce_sum(row_lengths, keepdims=True)
  dims = ragged_3d.shape[-1]
  outer_row_lengths = tf.cast(tf.fill(total_steps, dims), tf.int64)
  return tf.RaggedTensor.from_nested_row_lengths(
    tf.reshape(ragged_3d.flat_values, [-1]),
    nested_row_lengths=[row_lengths, outer_row_lengths])
```

Here's how it can be used:

```python
import tensorflow as tf

A_ids = tf.ragged.constant([[1, 2], [1, 0, 5]])
B_ids = tf.ragged.constant([[2, 4, 6, 9], [3]])
embedding_layer = tf.keras.layers.Embedding(10, 3)
A = embedding_layer(A_ids)  # A has shape [2, None, 3]
B = embedding_layer(B_ids)  # B has shape [2, None, 3]
C = tf.matmul(A, unset_outer_dimension(B), transpose_b=True)  # WORKS FINE NOW
```

However, I haven't checked the performance, it might be really slow.

**Comment by sushreebarsa** [comment]
@gadagashwini Was able to replicate the issue on colab using TF [v2.8.0 ](https://colab.research.google.com/gist/sushreebarsa/e8b58c31cb9de99641fd3322dec6c78d/tensorflow-issue-matmul-ragged.ipynb), [nightly](https://colab.research.google.com/gist/sushreebarsa/1737ae9a7d4b0e1acb5d1a852299477f/tensorflow-issue-matmul-ragged.ipynb#scrollTo=SXr2SghiO1Y_)(2.9.0.dev20220313), please find the attached gists for reference. Thanks!

**Comment by edloper** [comment]
Thanks for reporting this!  Fixed in abc6f6ec7869d846b33b3d8e5e33284aac1600dc.
